File: tools/content_tools/content_tools/validate_knowledge_repo.py
from __future__ import annotations

import argparse
import json
import re
from collections.abc import Mapping, Sequence
from pathlib import Path
from typing import Any

from content_tools import sop_parse


REQUIRED_SCAFFOLD_DIRS = [
    "content",
    "workflow-templates",
    "assistant-prompts",
    "assistant-process",
    "examples",
    "images",
    "indexes",
    "schemas",
    "scripts",
    "tests",
]
EXPECTED_TEMPLATE_SLUGS = [
    "book-of-the-week",
    "course",
    "maven-ll",
    "newsletter",
    "office-hours",
    "oss",
    "podcast",
    "social-media",
    "tax-report",
    "webinar",
    "workshop",
]
WORKFLOW_SCHEMA_REQUIRED_FIELDS = {
    "id",
    "type",
    "name",
    "schema_version",
    "trigger",
    "bundle_links",
    "phases",
    "tasks",
    "default_assignee",
    "source_document_ids",
}
TASK_REQUIRED_FIELDS = {
    "id",
    "name",
    "phase_id",
    "stage",
    "schedule",
    "default_assignee",
    "required_proofs",
    "required_links",
    "instruction_doc_id",
}
TARGET_PATH_RE = re.compile(r"^workflow-templates/[a-z0-9][a-z0-9-]*\.yaml$")
STABLE_ID_RE = re.compile(r"^task-template\.tasks\.[a-z0-9][a-z0-9-]*$")
RUNTIME_TYPE_RE = re.compile(r"^[a-z0-9][a-z0-9-]*$")


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate the dataops-knowledge migration scaffold.")
    parser.add_argument("--repo-root", default=Path.cwd(), type=Path)
    parser.add_argument("--scaffold-root", default=Path("templates/dataops-knowledge"), type=Path)
    args = parser.parse_args()

    violations = validate(args.repo_root, args.scaffold_root)
    if violations:
        print("Knowledge repository scaffold validation failed:")
        for violation in violations:
            print(f"- {violation}")
        return 1

    print("Knowledge repository scaffold validation passed.")
    return 0


def validate(repo_root: Path, scaffold_root: Path | str = "templates/dataops-knowledge") -> list[str]:
    repo_root = repo_root.resolve()
    scaffold_path = _resolve_repo_path(repo_root, scaffold_root)
    violations: list[str] = []

    violations.extend(validate_scaffold_dirs(repo_root, scaffold_path))

    schema_path = scaffold_path / "schemas" / "workflow-template.schema.json"
    schema = _load_json(schema_path, repo_root, violations, "workflow-template schema")
    if isinstance(schema, dict):
        violations.extend(validate_workflow_schema_shape(schema, repo_root, schema_path))

    manifest_path = scaffold_path / "indexes" / "workflow-template-migration-manifest.json"
    manifest = _load_json(manifest_path, repo_root, violations, "workflow-template migration manifest")
    if isinstance(manifest, dict):
        violations.extend(validate_manifest(repo_root, manifest_path, manifest))

    return violations


def validate_scaffold_dirs(repo_root: Path, scaffold_root: Path) -> list[str]:
    violations: list[str] = []
    if not scaffold_root.exists():
        return [f"{_repo_path(repo_root, scaffold_root)}: scaffold directory is required"]
    for relative_dir in REQUIRED_SCAFFOLD_DIRS:
        path = scaffold_root / relative_dir
        if not path.is_dir():
            violations.append(f"{_repo_path(repo_root, path)}: required scaffold directory is missing")
    return violations


def validate_workflow_schema_shape(schema: Mapping[str, Any], repo_root: Path, schema_path: Path) -> list[str]:
    source = _repo_path(repo_root, schema_path)
    violations: list[str] = []
    if schema.get("type") != "object":
        violations.append(f"{source}: schema root type must be object")
    if schema.get("additionalProperties") is not False:
        violations.append(f"{source}: schema root must set additionalProperties to false")

    required = set(_strings(schema.get("required")))
    missing_required = sorted(WORKFLOW_SCHEMA_REQUIRED_FIELDS - required)
    for field in missing_required:
        violations.append(f"{source}: schema required fields missing {field!r}")

    properties = schema.get("properties")
    if not isinstance(properties, dict):
        return violations + [f"{source}: schema properties must be an object"]
    for field in sorted(WORKFLOW_SCHEMA_REQUIRED_FIELDS):
        if field not in properties:
            violations.append(f"{source}: schema properties missing {field!r}")

    defs = schema.get("$defs")
    if not isinstance(defs, dict):
        return violations + [f"{source}: schema $defs must be an object"]
    task_schema = defs.get("task")
    if not isinstance(task_schema, dict):
        return violations + [f"{source}: schema $defs.task is required"]
    task_required = set(_strings(task_schema.get("required")))
    for field in sorted(TASK_REQUIRED_FIELDS - task_required):
        violations.append(f"{source}: task schema required fields missing {field!r}")

    return violations


def validate_manifest(repo_root: Path, manifest_path: Path, manifest: Mapping[str, Any]) -> list[str]:
    source = _repo_path(repo_root, manifest_path)
    violations: list[str] = []
    if manifest.get("schema_version") != 1:
        violations.append(f"{source}: schema_version must be 1")
    if manifest.get("source_root") != "content/tasks/templates":
        violations.append(f"{source}: source_root must be 'content/tasks/templates'")
    if manifest.get("target_root") != "workflow-templates":
        violations.append(f"{source}: target_root must be 'workflow-templates'")

    entries = manifest.get("templates")
    if not isinstance(entries, list):
        return violations + [f"{source}: templates must be a list"]

    current_template_paths = sorted((repo_root / "content" / "tasks" / "templates").glob("*.md"))
    current_sources = {_repo_path(repo_root, path) for path in current_template_paths}
    expected_sources = {f"content/tasks/templates/{slug}.md" for slug in EXPECTED_TEMPLATE_SLUGS}
    if current_sources != expected_sources:
        missing_current = sorted(expected_sources - current_sources)
        extra_current = sorted(current_sources - expected_sources)
        for path in missing_current:
            violations.append(f"{path}: expected current Markdown template file is missing")
        for path in extra_current:
            violations.append(f"{path}: unexpected current Markdown template file is present")

    seen_ids: dict[str, int] = {}
    seen_sources: dict[str, int] = {}
    seen_targets: dict[str, int] = {}
    mapped_sources: set[str] = set()

    for index, entry in enumerate(entries):
        entry_label = f"{source}: templates[{index}]"
        if not isinstance(entry, dict):
            violations.append(f"{entry_label}: entry must be an object")
            continue

        stable_id = _required_string(entry, "stable_id", entry_label, violations)
        runtime_type = _required_string(entry, "runtime_type", entry_label, violations)
        source_path = _required_string(entry, "source_path", entry_label, violations)
        target_path = _required_string(entry, "target_path", entry_label, violations)
        if not all([stable_id, runtime_type, source_path, target_path]):
            continue

        if not STABLE_ID_RE.fullmatch(stable_id):
            violations.append(f"{entry_label}: stable_id must match task-template.tasks.<slug>, got {stable_id!r}")
        if not RUNTIME_TYPE_RE.fullmatch(runtime_type):
            violations.append(f"{entry_label}: runtime_type must be a slug, got {runtime_type!r}")
        if source_path not in expected_sources:
            violations.append(f"{entry_label}: source_path must be one of the current task templates, got {source_path!r}")
        if not TARGET_PATH_RE.fullmatch(target_path):
            violations.append(f"{entry_label}: target_path must match workflow-templates/*.yaml, got {target_path!r}")

        _record_duplicate(seen_ids, stable_id, index, entry_label, "stable_id", violations)
        _record_duplicate(seen_sources, source_path, index, entry_label, "source_path", violations)
        _record_duplicate(seen_targets, target_path, index, entry_label, "target_path", violations)
        mapped_sources.add(source_path)

        template_path = repo_root / source_path
        if not template_path.is_file():
            violations.append(f"{entry_label}: source Markdown template is missing: {source_path}")
            continue
        violations.extend(validate_current_template_frontmatter(repo_root, template_path, stable_id, entry_label))

    missing_mappings = sorted(expected_sources - mapped_sources)
    extra_mappings = sorted(mapped_sources - expected_sources)
    for source_path in missing_mappings:
        violations.append(f"{source}: missing migration mapping for {source_path}")
    for source_path in extra_mappings:
        violations.append(f"{source}: unexpected migration mapping for {source_path}")
    if len(entries) != len(EXPECTED_TEMPLATE_SLUGS):
        violations.append(f"{source}: expected {len(EXPECTED_TEMPLATE_SLUGS)} template mappings, found {len(entries)}")

    return violations


def validate_current_template_frontmatter(repo_root: Path, template_path: Path, expected_id: str, label: str) -> list[str]:
    violations: list[str] = []
    text = template_path.read_text(encoding="utf-8", errors="replace")
    raw, _body = sop_parse.split_frontmatter(text)
    metadata = sop_parse.parse_frontmatter(raw) if raw else {}
    actual_id = str(metadata.get("id", "")).strip()
    if not actual_id:
        violations.append(f"{label}: {_repo_path(repo_root, template_path)} frontmatter id is required")
    elif actual_id != expected_id:
        violations.append(
            f"{label}: {_repo_path(repo_root, template_path)} frontmatter id {actual_id!r} does not match stable_id {expected_id!r}"
        )
    if metadata.get("doc_type") != "task-template":
        violations.append(
            f"{label}: {_repo_path(repo_root, template_path)} frontmatter doc_type must be task-template, got {metadata.get('doc_type')!r}"
        )
    return violations


def validate_workflow_template_document(document: Mapping[str, Any], schema: Mapping[str, Any]) -> list[str]:
    """Validate a workflow-template document against the supported JSON Schema subset."""
    return _validate_schema_node(document, schema, schema, "$")


def _validate_schema_node(value: Any, schema: Mapping[str, Any], root_schema: Mapping[str, Any], path: str) -> list[str]:
    if "$ref" in schema:
        target = _resolve_schema_ref(root_schema, str(schema["$ref"]))
        if target is None:
            return [f"{path}: unresolved schema reference {schema['$ref']!r}"]
        return _validate_schema_node(value, target, root_schema, path)

    violations: list[str] = []
    expected_type = schema.get("type")
    if expected_type and not _matches_json_type(value, str(expected_type)):
        return [f"{path}: expected {expected_type}, got {_json_type_name(value)}"]

    if "const" in schema and value != schema["const"]:
        violations.append(f"{path}: expected const {schema['const']!r}, got {value!r}")
    if "enum" in schema and value not in schema["enum"]:
        violations.append(f"{path}: expected one of {schema['enum']!r}, got {value!r}")
    if isinstance(value, str):
        min_length = schema.get("minLength")
        if isinstance(min_length, int) and len(value) < min_length:
            violations.append(f"{path}: string is shorter than minLength {min_length}")
        pattern = schema.get("pattern")
        if isinstance(pattern, str) and not re.fullmatch(pattern, value):
            violations.append(f"{path}: value {value!r} does not match pattern {pattern!r}")

    if isinstance(value, dict):
        required = _strings(schema.get("required"))
        for field in required:
            if field not in value:
                violations.append(f"{path}: required field {field!r} is missing")
        properties = schema.get("properties") if isinstance(schema.get("properties"), dict) else {}
        if schema.get("additionalProperties") is False:
            for field in value:
                if field not in properties:
                    violations.append(f"{path}.{field}: additional property is not allowed")
        for field, field_schema in properties.items():
            if field in value and isinstance(field_schema, dict):
                violations.extend(_validate_schema_node(value[field], field_schema, root_schema, f"{path}.{field}"))
        any_of = schema.get("anyOf")
        if isinstance(any_of, list) and any_of:
            if not any(
                isinstance(candidate, dict) and not _validate_schema_node(value, candidate, root_schema, path)
                for candidate in any_of
            ):
                violations.append(f"{path}: value does not match any required schema option")

    if isinstance(value, list):
        min_items = schema.get("minItems")
        if isinstance(min_items, int) and len(value) < min_items:
            violations.append(f"{path}: array has fewer than {min_items} items")
        if schema.get("uniqueItems") is True and len({_json_key(item) for item in value}) != len(value):
            violations.append(f"{path}: array items must be unique")
        item_schema = schema.get("items")
        if isinstance(item_schema, dict):
            for index, item in enumerate(value):
                violations.extend(_validate_schema_node(item, item_schema, root_schema, f"{path}[{index}]"))

    return violations


def _load_json(path: Path, repo_root: Path, violations: list[str], label: str) -> Any:
    if not path.is_file():
        violations.append(f"{_repo_path(repo_root, path)}: {label} file is required")
        return None
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        violations.append(f"{_repo_path(repo_root, path)}: invalid JSON: {exc.msg} at line {exc.lineno}, column {exc.colno}")
        return None


def _record_duplicate(
    seen: dict[str, int],
    value: str,
    index: int,
    entry_label: str,
    field: str,
    violations: list[str],
) -> None:
    if value in seen:
        violations.append(f"{entry_label}: duplicate {field} {value!r}; first seen at templates[{seen[value]}]")
    else:
        seen[value] = index


def _required_string(entry: Mapping[str, Any], key: str, label: str, violations: list[str]) -> str:
    value = entry.get(key)
    if not isinstance(value, str) or not value.strip():
        violations.append(f"{label}: {key} is required")
        return ""
    return value.strip()


def _resolve_schema_ref(root_schema: Mapping[str, Any], ref: str) -> Mapping[str, Any] | None:
    if not ref.startswith("#/"):
        return None
    node: Any = root_schema
    for part in ref[2:].split("/"):
        if not isinstance(node, dict):
            return None
        node = node.get(part)
    return node if isinstance(node, dict) else None


def _matches_json_type(value: Any, expected_type: str) -> bool:
    if expected_type == "object":
        return isinstance(value, dict)
    if expected_type == "array":
        return isinstance(value, list)
    if expected_type == "string":
        return isinstance(value, str)
    if expected_type == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if expected_type == "boolean":
        return isinstance(value, bool)
    return True


def _json_type_name(value: Any) -> str:
    if isinstance(value, dict):
        return "object"
    if isinstance(value, list):
        return "array"
    if isinstance(value, str):
        return "string"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, int):
        return "integer"
    if value is None:
        return "null"
    return type(value).__name__


def _json_key(value: Any) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"))


def _strings(value: Any) -> list[str]:
    if isinstance(value, str):
        return [value]
    if isinstance(value, Sequence) and not isinstance(value, (str, bytes)):
        return [item for item in value if isinstance(item, str)]
    return []


def _resolve_repo_path(repo_root: Path, path: Path | str) -> Path:
    path = Path(path)
    if path.is_absolute():
        return path.resolve()
    return (repo_root / path).resolve()


def _repo_path(repo_root: Path, path: Path) -> str:
    try:
        return path.resolve().relative_to(repo_root).as_posix()
    except ValueError:
        return path.as_posix()


if __name__ == "__main__":
    raise SystemExit(main())

File: backend/scripts/validate-knowledge-repo.ts
#!/usr/bin/env node

/**
 * Validates the dataops-knowledge migration scaffold: the required directory
 * layout, the workflow-template JSON Schema shape, and the migration manifest
 * that maps each current Markdown task template to its future YAML target.
 *
 * Usage:
 *   npx tsx scripts/validate-knowledge-repo.ts [--repo-root <path>] [--scaffold-root <path>]
 */

import { existsSync, readFileSync, readdirSync, statSync } from 'node:fs';
import { isAbsolute, join, relative, resolve } from 'node:path';

import { parseFrontmatter, splitFrontmatter } from '../src/docs/sop';

const REQUIRED_SCAFFOLD_DIRS = [
  'content', 'workflow-templates', 'assistant-prompts', 'assistant-process',
  'examples', 'images', 'indexes', 'schemas', 'scripts', 'tests',
];
const EXPECTED_TEMPLATE_SLUGS = [
  'book-of-the-week', 'course', 'maven-ll', 'newsletter', 'office-hours',
  'oss', 'podcast', 'social-media', 'tax-report', 'webinar', 'workshop',
];
const WORKFLOW_SCHEMA_REQUIRED_FIELDS = [
  'id', 'type', 'name', 'schema_version', 'trigger', 'card_links', 'phases',
  'tasks', 'default_assignee', 'source_document_ids',
];
const TASK_REQUIRED_FIELDS = [
  'id', 'name', 'phase_id', 'stage', 'schedule', 'default_assignee',
  'required_proofs', 'required_links', 'instruction_doc_id',
];
const TARGET_PATH_RE = /^workflow-templates\/[a-z0-9][a-z0-9-]*\.yaml$/;
const STABLE_ID_RE = /^task-template\.tasks\.[a-z0-9][a-z0-9-]*$/;
const RUNTIME_TYPE_RE = /^[a-z0-9][a-z0-9-]*$/;

type Json = unknown;
type JsonObject = Record<string, Json>;

const isObject = (value: Json): value is JsonObject =>
  typeof value === 'object' && value !== null && !Array.isArray(value);

export function validate(repoRoot: string, scaffoldRoot = 'templates/dataops-knowledge'): string[] {
  const root = resolve(repoRoot);
  const scaffold = resolve(isAbsolute(scaffoldRoot) ? scaffoldRoot : join(root, scaffoldRoot));
  const violations: string[] = [];

  violations.push(...validateScaffoldDirs(root, scaffold));

  const schemaPath = join(scaffold, 'schemas', 'workflow-template.schema.json');
  const schema = loadJson(schemaPath, root, violations, 'workflow-template schema');
  if (isObject(schema)) violations.push(...validateWorkflowSchemaShape(schema, root, schemaPath));

  const manifestPath = join(scaffold, 'indexes', 'workflow-template-migration-manifest.json');
  const manifest = loadJson(manifestPath, root, violations, 'workflow-template migration manifest');
  if (isObject(manifest)) violations.push(...validateManifest(root, manifestPath, manifest));

  return violations;
}

function validateScaffoldDirs(repoRoot: string, scaffoldRoot: string): string[] {
  if (!existsSync(scaffoldRoot)) {
    return [`${repoPath(repoRoot, scaffoldRoot)}: scaffold directory is required`];
  }
  const violations: string[] = [];
  for (const dir of REQUIRED_SCAFFOLD_DIRS) {
    const path = join(scaffoldRoot, dir);
    if (!existsSync(path) || !statSync(path).isDirectory()) {
      violations.push(`${repoPath(repoRoot, path)}: required scaffold directory is missing`);
    }
  }
  return violations;
}

function validateWorkflowSchemaShape(schema: JsonObject, repoRoot: string, schemaPath: string): string[] {
  const source = repoPath(repoRoot, schemaPath);
  const violations: string[] = [];
  if (schema.type !== 'object') violations.push(`${source}: schema root type must be object`);
  if (schema.additionalProperties !== false) {
    violations.push(`${source}: schema root must set additionalProperties to false`);
  }

  const required = new Set(asStrings(schema.required));
  for (const field of [...WORKFLOW_SCHEMA_REQUIRED_FIELDS].filter((f) => !required.has(f)).sort()) {
    violations.push(`${source}: schema required fields missing '${field}'`);
  }

  if (!isObject(schema.properties)) {
    violations.push(`${source}: schema properties must be an object`);
    return violations;
  }
  const properties = schema.properties;
  for (const field of [...WORKFLOW_SCHEMA_REQUIRED_FIELDS].sort()) {
    if (!(field in properties)) violations.push(`${source}: schema properties missing '${field}'`);
  }

  if (!isObject(schema.$defs)) {
    violations.push(`${source}: schema $defs must be an object`);
    return violations;
  }
  const taskSchema = schema.$defs.task;
  if (!isObject(taskSchema)) {
    violations.push(`${source}: schema $defs.task is required`);
    return violations;
  }
  const taskRequired = new Set(asStrings(taskSchema.required));
  for (const field of [...TASK_REQUIRED_FIELDS].filter((f) => !taskRequired.has(f)).sort()) {
    violations.push(`${source}: task schema required fields missing '${field}'`);
  }
  return violations;
}

function validateManifest(repoRoot: string, manifestPath: string, manifest: JsonObject): string[] {
  const source = repoPath(repoRoot, manifestPath);
  const violations: string[] = [];
  if (manifest.schema_version !== 1) violations.push(`${source}: schema_version must be 1`);
  if (manifest.source_root !== 'content/tasks/templates') {
    violations.push(`${source}: source_root must be 'content/tasks/templates'`);
  }
  if (manifest.target_root !== 'workflow-templates') {
    violations.push(`${source}: target_root must be 'workflow-templates'`);
  }

  const entries = manifest.templates;
  if (!Array.isArray(entries)) {
    violations.push(`${source}: templates must be a list`);
    return violations;
  }

  const templatesDir = join(repoRoot, 'content', 'tasks', 'templates');
  const currentSources = new Set(
    (existsSync(templatesDir) ? readdirSync(templatesDir) : [])
      .filter((name) => name.endsWith('.md'))
      .map((name) => repoPath(repoRoot, join(templatesDir, name))),
  );
  const expectedSources = new Set(EXPECTED_TEMPLATE_SLUGS.map((slug) => `content/tasks/templates/${slug}.md`));
  if (!setsEqual(currentSources, expectedSources)) {
    for (const path of difference(expectedSources, currentSources)) {
      violations.push(`${path}: expected current Markdown template file is missing`);
    }
    for (const path of difference(currentSources, expectedSources)) {
      violations.push(`${path}: unexpected current Markdown template file is present`);
    }
  }

  const seenIds = new Map<string, number>();
  const seenSources = new Map<string, number>();
  const seenTargets = new Map<string, number>();
  const mappedSources = new Set<string>();

  entries.forEach((entry, index) => {
    const label = `${source}: templates[${index}]`;
    if (!isObject(entry)) {
      violations.push(`${label}: entry must be an object`);
      return;
    }

    const stableId = requiredString(entry, 'stable_id', label, violations);
    const runtimeType = requiredString(entry, 'runtime_type', label, violations);
    const sourcePath = requiredString(entry, 'source_path', label, violations);
    const targetPath = requiredString(entry, 'target_path', label, violations);
    if (!stableId || !runtimeType || !sourcePath || !targetPath) return;

    if (!STABLE_ID_RE.test(stableId)) {
      violations.push(`${label}: stable_id must match task-template.tasks.<slug>, got '${stableId}'`);
    }
    if (!RUNTIME_TYPE_RE.test(runtimeType)) {
      violations.push(`${label}: runtime_type must be a slug, got '${runtimeType}'`);
    }
    if (!expectedSources.has(sourcePath)) {
      violations.push(`${label}: source_path must be one of the current task templates, got '${sourcePath}'`);
    }
    if (!TARGET_PATH_RE.test(targetPath)) {
      violations.push(`${label}: target_path must match workflow-templates/*.yaml, got '${targetPath}'`);
    }

    recordDuplicate(seenIds, stableId, index, label, 'stable_id', violations);
    recordDuplicate(seenSources, sourcePath, index, label, 'source_path', violations);
    recordDuplicate(seenTargets, targetPath, index, label, 'target_path', violations);
    mappedSources.add(sourcePath);

    const templatePath = join(repoRoot, sourcePath);
    if (!existsSync(templatePath) || !statSync(templatePath).isFile()) {
      violations.push(`${label}: source Markdown template is missing: ${sourcePath}`);
      return;
    }
    violations.push(...validateCurrentTemplateFrontmatter(repoRoot, templatePath, stableId, label));
  });

  for (const sourcePath of difference(expectedSources, mappedSources)) {
    violations.push(`${source}: missing migration mapping for ${sourcePath}`);
  }
  for (const sourcePath of difference(mappedSources, expectedSources)) {
    violations.push(`${source}: unexpected migration mapping for ${sourcePath}`);
  }
  if (entries.length !== EXPECTED_TEMPLATE_SLUGS.length) {
    violations.push(`${source}: expected ${EXPECTED_TEMPLATE_SLUGS.length} template mappings, found ${entries.length}`);
  }

  return violations;
}

function validateCurrentTemplateFrontmatter(
  repoRoot: string,
  templatePath: string,
  expectedId: string,
  label: string,
): string[] {
  const violations: string[] = [];
  const [raw] = splitFrontmatter(readFileSync(templatePath, 'utf8'));
  const metadata = (raw ? parseFrontmatter(raw) : {}) as Record<string, unknown>;
  const path = repoPath(repoRoot, templatePath);
  const actualId = String(metadata.id ?? '').trim();
  if (!actualId) violations.push(`${label}: ${path} frontmatter id is required`);
  else if (actualId !== expectedId) {
    violations.push(`${label}: ${path} frontmatter id '${actualId}' does not match stable_id '${expectedId}'`);
  }
  if (metadata.doc_type !== 'task-template') {
    const actual = metadata.doc_type === undefined ? 'None' : `'${String(metadata.doc_type)}'`;
    violations.push(`${label}: ${path} frontmatter doc_type must be task-template, got ${actual}`);
  }
  return violations;
}

function loadJson(path: string, repoRoot: string, violations: string[], label: string): Json {
  if (!existsSync(path) || !statSync(path).isFile()) {
    violations.push(`${repoPath(repoRoot, path)}: ${label} file is required`);
    return null;
  }
  try {
    return JSON.parse(readFileSync(path, 'utf8'));
  } catch (error) {
    violations.push(`${repoPath(repoRoot, path)}: invalid JSON: ${(error as Error).message}`);
    return null;
  }
}

function recordDuplicate(
  seen: Map<string, number>,
  value: string,
  index: number,
  label: string,
  field: string,
  violations: string[],
): void {
  const first = seen.get(value);
  if (first !== undefined) {
    violations.push(`${label}: duplicate ${field} '${value}'; first seen at templates[${first}]`);
  } else {
    seen.set(value, index);
  }
}

function requiredString(entry: JsonObject, key: string, label: string, violations: string[]): string {
  const value = entry[key];
  if (typeof value !== 'string' || !value.trim()) {
    violations.push(`${label}: ${key} is required`);
    return '';
  }
  return value.trim();
}

function asStrings(value: Json): string[] {
  if (typeof value === 'string') return [value];
  if (Array.isArray(value)) return value.filter((item): item is string => typeof item === 'string');
  return [];
}

function difference(a: Set<string>, b: Set<string>): string[] {
  return [...a].filter((item) => !b.has(item)).sort();
}

function setsEqual(a: Set<string>, b: Set<string>): boolean {
  return a.size === b.size && [...a].every((item) => b.has(item));
}

function repoPath(repoRoot: string, path: string): string {
  const rel = relative(repoRoot, resolve(path));
  if (rel.startsWith('..')) return resolve(path).split(/[\\/]/).join('/');
  return rel.split(/[\\/]/).join('/');
}

function readFlag(name: string, fallback: string): string {
  // Last occurrence wins, so `npm run <script> -- --flag value` overrides the
  // value baked into the npm script rather than being silently ignored.
  const index = process.argv.lastIndexOf(name);
  if (index === -1) return fallback;
  const value = process.argv[index + 1];
  if (!value || value.startsWith('--')) throw new Error(`${name} requires a value`);
  return value;
}

if (require.main === module) {
  const violations = validate(readFlag('--repo-root', process.cwd()), readFlag('--scaffold-root', 'templates/dataops-knowledge'));
  if (violations.length > 0) {
    console.error('Knowledge repository scaffold validation failed:');
    violations.forEach((violation) => console.error(`- ${violation}`));
    process.exit(1);
  }
  console.log('Knowledge repository scaffold validation passed.');
}
